frontend/src/app/modules/user/submissions-chart/submissions-chart.component.ts

```diff
@@ -0,0 +1,142 @@
+
+    ngOnInit(): void {
+        this.updateChartData();
+    }
+
+    ngOnChanges(): void {
+        this.updateChartData();
+    }
+
+    xAxisTickFormatting(value: string) {
+        const columnData = value.split(';');
+
+        return columnData[0];
+    }
+
+    private getSubmissionsCount(date: Date) {
+        return this.solutions.reduce((result, solution) => {
+            if (solution.submittedAt && moment(solution.submittedAt).startOf('day').isSame(date)) {
+                result++;
+            }
+
+            return result;
+        }, 0);
+    }
+
+    private updateChartData() {
+        this.chartData = [];
+        this.totalSubmissions = 0;
+        this.totalActiveDays = 0;
+        this.maxStreak = 0;
+        let currentStreak = 0;
+
+        const firstDay = moment().subtract(1, 'year').add(1, 'day').startOf('day')
+            .toDate();
+        let seriesMonth = moment(firstDay).startOf('month');
+        const seriesWeek = moment(firstDay).startOf('week');
+        const lastWeek = moment().startOf('week');
+        let monthChartData: WeekChartData[] = [];
+        let columnId = 0;
+
+        while (seriesWeek <= lastWeek) {
+            columnId++;
+            const seriesData: DayChartData[] = [];
+
+            for (let i = 6; i >= 0; i--) {
+                const weekDate = seriesWeek.clone().add(i, 'day');
+                const date = weekDate.toDate();
+                const name = weekDate.format('ddd');
+                let value = -1; //not set
+
+                if (weekDate.clone().startOf('month').isSame(seriesMonth)) {
+                    value = this.getSubmissionsCount(date);
+                    this.totalSubmissions += value;
+                    if (value > 0) {
+                        currentStreak++;
+                        this.totalActiveDays++;
+                        this.maxStreak = Math.max(this.maxStreak, currentStreak);
+                    } else {
+                        currentStreak = 0;
+                    }
+                }
+
+                seriesData.push({ date, name, value });
+            }
+
+            const weekChartData: WeekChartData = {
+                name: `;${columnId}`,
+                series: seriesData,
+            };
+
+            this.chartData.push(weekChartData);
+            if (seriesMonth.isSame(seriesWeek, 'month')
+                || seriesMonth.isSame(seriesWeek.clone().endOf('week'), 'month')) {
+                monthChartData.push(weekChartData);
+            }
+
+            const nextWeekMonth = seriesWeek.clone().add(1, 'week').startOf('month');
+
+            if (seriesMonth.isSame(nextWeekMonth)) {
+                seriesWeek.add(1, 'week');
+            } else {
+                this.setMonthTitle(monthChartData, seriesMonth.format('MMM'));
+                seriesMonth = nextWeekMonth.clone();
+                monthChartData = [];
+            }
+        }
+        this.setMonthTitle(monthChartData, seriesMonth.format('MMM'));
```

Comment:
too long methods
Split it to smaller, it will make it more redable

Reply:
refactored